### PR TITLE
fix: auth expiration and error handling

### LINKED ISSUE #565: Signing out of the CCloud web UI after signing in through the extension results in a bad/mixed auth state
When our auth polling notices [an error](https://github.com/confluentinc/vscode/blob/v0.20.x/src/sidecar/authStatusPolling.ts#L272-L283) (not expiration-related), it shows a `Log in to Confluent Cloud` button:

image

When clicked, it calls into the CCloud auth provider's `createSession()` method, looks whether or not we have an [existing CCloud connection](https://github.com/confluentinc/vscode/blob/v0.20.x/src/authProvider.ts#L127-L132), then tries to [open a browser tab](https://github.com/confluentinc/vscode/blob/v0.20.x/src/authProvider.ts#L297) with the existing sign-in URI.

Something in this `browserAuthFlow()` is silently breaking, as the browser tab doesn't open up as intended. This then means the extension is stuck knowing about an existing unusable CCloud auth session, the sidebar doesn't clear out, and it's generally a bad experience.

At the very least, we need to do two things:
- clear the auth session from the VS Code authentication perspective (but not delete the sidecar connection)
- fix whatever is breaking while trying to open the browser tab to reauth

(More details TBD)

### PR DESCRIPTION
**Pull request by shouples**
## Summary of Changes

Closes #565 and #500.

This PR fixes two fairly big issues with CCloud authentication/connection management:
- When we get `errors` from a non-`INVALID_TOKEN` status, we weren't invalidating the (VS Code) auth session. This meant all of the sidebar content would persist and appear to the user that they could still interact with their CCloud resources, when that really wasn't the case. ([see fix](https://github.com/confluentinc/vscode/pull/566#discussion_r1835184182))
- The error notification that appeared from an auth-errors scenario would show a "Log in to Confluent Cloud" button that made a call to our auth provider, but it wouldn't actually open the browser due to a flaw in the provider's `getSessions()` logic:
  - When the button is clicked (and `authentication.getSession(AUTH_PROVIDER_ID, [], { createIfNone: true })` is called), it goes through the provider's `getSessions()` method, where the sidecar says we have a connection (no matter what auth status). Our auth provider considered it usable as a VS Code authentication session and therefore never made it to `createSession()` and never opened the browser tab. ([see fix](https://github.com/confluentinc/vscode/pull/566#discussion_r1835184269))
    - Due to connections with `NO_TOKEN`/`FAILED` auth status being accepted, when we converted the connection to a VS Code auth session, we would fall back to using the `unk username` values.

Also fixes a few minor issues:
- We no longer save the auth status (`VALID_TOKEN`, `FAILED`, etc) as the VS Code authentication session's `accessToken`, which means we should no longer see the "Grant access to _____" notifications in the Accounts menu anymore. This is because the session will either exist, or it won't; there will be no same-session-different-token mismatching to confuse VS Code.
- `ccloudAuthSessionInvalidated` no longer tells the auth provider to delete the CCloud connection entirely; that was leftover from before we had the `FAILED` auth status as a means of "resetting" the connection. Instead, it just removes the session from the auth provider's internal tracking, which primarily just prompts the user to sign in again.
- The ordering of the auth poller's check for expiration and errors is now flipped, so expired CCloud auth will float the notification and exit earlier and not show double "Confluent Cloud authentication expired at ____" + "Error authenticating with Confluent Cloud." notifications

## Any additional details or context that should be provided?

I've introduced a new `logResponseError()` helper function to capture a lot of the `if (error instanceof ResponseError)` logging dance we've done in a number of other places, because we were seeing response errors during connection creation/deletion while debugging the primary issues mentioned above earlier today. I plan on making some follow-on PRs to use it more and reduce some boilerplate code while increasing our visibility in error cases (if the user has telemetry enabled).

## Pull request checklist

Please check if your PR fulfills the following (if applicable):

##### Tests

- [x] Added new
- [x] Updated existing
- [ ] Deleted existing

##### Other

- [ ] All new disposables (event listeners, views, channels, etc.) collected as  for eventual cleanup?

- [x] Does anything in this PR need to be mentioned in the user-facing [CHANGELOG](https://github.com/confluentinc/vscode/blob/main/CHANGELOG.md) or [README](https://github.com/confluentinc/vscode/blob/main/public/README.md)?
- [x] Have you validated this change locally by [packaging](https://github.com/confluentinc/vscode/blob/main/README.md#packaging-steps) and installing the extension `.vsix` file?
  ```shell
  gulp clicktest
  ```

### CODE REVIEW

File: src/authn/ccloudPolling.ts [hunk 1]
```diff
@@ -243,16 +251,17 @@ async function handleExpiredAuth(expirationString: string) {
...
+/** Check the {@link Connection} for any auth-related errors and prompt the user to reauthenticate. */
 export function checkAuthErrors(connection: Connection) {
   const errors: AuthErrors | undefined = connection.status.authentication.errors;
   if (!errors) {
     return;
   }
 
+  // tell the auth provider that the session is invalid so it can prompt the user to log in again
+  // and clear out the sidebar resources (but not delete the connection)
+  ccloudAuthSessionInvalidated.fire();
```
Comment (shouples) [inline src/authn/ccloudPolling.ts:263]: Fix for first bullet point in PR description

File: src/authn/ccloudProvider.ts [hunk 2]
```diff
@@ -134,15 +150,29 @@ export class ConfluentCloudAuthProvider implements vscode.AuthenticationProvider
...
 
+    if (connection && ["NO_TOKEN", "FAILED"].includes(connection.status.authentication.status)) {
+      // the connection is unusable, so the auth provider needs to act like there isn't actually a
+      // connection and the user needs to sign in
+      logger.debug("getSessions() found connection with unusable status", {
+        status: connection.status.authentication.status,
+      });
+      connection = null;
+    }
+
```
Comment (shouples) [inline src/authn/ccloudProvider.ts:175]: Fix for second bullet point in PR description

File: src/authn/ccloudProvider.ts [hunk 3]
```diff
@@ -297,11 +324,13 @@ export class ConfluentCloudAuthProvider implements vscode.AuthenticationProvider
     });
 
     // if any other part of the extension notices that our current CCloud connection transitions from
-    // VALID_TOKEN to INVALID_TOKEN/NO_TOKEN, we need to remove the session and stop polling
+    // VALID_TOKEN to FAILED/NO_TOKEN, we need to remove the session and stop polling
     const ccloudAuthSessionInvalidatedSub: vscode.Disposable = ccloudAuthSessionInvalidated.event(
       async () => {
         logger.debug("ccloudAuthSessionInvalidated event fired");
-        await this.removeSession(CCLOUD_CONNECTION_ID);
+        // don't delete the actual CCloud connection, just remove it from the authentication provider
```
Comment (jlrobins) [inline src/authn/ccloudProvider.ts:331]: Would love more comment here, as to *why* deleting is bad, but "remove it from the auth provider" is good and proper.

Or, perhaps rather talk positively about what we do need to do any why, and don't talk at all about what we don't need to do.

File: src/errors.ts [hunk 4]
```diff
@@ -1,6 +1,60 @@
...
+
+/** Combined ResponseError type from our OpenAPI spec generated client code. */
+type ResponseError =
+  | SidecarResponseError
+  | KafkaResponseError
+  | SchemaRegistryResponseError
+  | DockerResponseError;
+
+/** Log the possibly-ResponseError and any additional information, and optionally send to Sentry. */
+export async function logResponseError(
```
Comment (jlrobins) [inline src/errors.ts:26]: Thank you very, Very, VERY much.

File: src/authn/ccloudPolling.ts [hunk 5]
```diff
@@ -92,26 +90,34 @@ export async function watchCCloudConnectionStatus(): Promise<void> {
...
   }
 
+  // if the auth status is still valid, but it's within {MINUTES_UNTIL_REAUTH_WARNING}min of expiring,
+  // warn the user to reauth; also handle if the session has already expired
+  const sessionExpired: boolean = await checkAuthExpiration(connection);
+  if (sessionExpired) {
+    // don't bother looking for errors if the session has already expired, the result is going to be
+    // the same: the user needs to reauthenticate
+    return;
+  }
```
Comment (flippingbits) [inline src/authn/ccloudPolling.ts:100]: I feel the extension should only consider the auth status for determining if the session has expired. In some cases, sessions expire before `connection.status.authentication.requires_authentication_at`, e.g., when the user has signed out from the CCloud UI. I'd suggest to use `connection.status.authentication.requires_authentication_at` only for determining if we need to warn the user about the soon-to-expire session. WDYT?

Comment (flippingbits) [reply src/authn/ccloudPolling.ts:99]: Got it. Thanks for the clarification!

File: src/authn/ccloudProvider.ts (hunk 3 above)
Comment (shouples) [reply src/authn/ccloudProvider.ts:366]: Added a bit more in https://github.com/confluentinc/vscode/pull/566/commits/cd11995dc652911159f4dfb4052dc0241318129a and also https://github.com/confluentinc/vscode/pull/566/commits/1de3a30068e4785e0a8c7a38304c05bb98ba06a3

**Review by jlrobins — APPROVED** [review]

**Pull request merged**